File: mobile/lib/screens/chat/reply_codec.dart
import 'dart:convert';

class ReplyDecoded {
  final Map<String, dynamic>? reply;
  final String body;
  const ReplyDecoded({required this.reply, required this.body});
}

ReplyDecoded decodeReply(String raw) {
  const prefix = '@@reply:';
  if (!raw.startsWith(prefix)) {
    return ReplyDecoded(reply: null, body: raw);
  }
  final end = raw.indexOf('@@', prefix.length);
  if (end == -1) return ReplyDecoded(reply: null, body: raw);
  final jsonPart = raw.substring(prefix.length, end);
  var body = raw.substring(end + 2);
  if (body.startsWith('\n')) body = body.substring(1);
  try {
    final obj = jsonDecode(jsonPart);
    if (obj is Map<String, dynamic>) {
      return ReplyDecoded(reply: obj, body: body);
    }
    return ReplyDecoded(reply: null, body: raw);
  } catch (_) {
    return ReplyDecoded(reply: null, body: raw);
  }
}

String stripReply(String raw) => decodeReply(raw).body;

String encodeReply({
  required String body,
  required int replyId,
  required String replySender,
  required String replySnippet,
}) {
  final meta = <String, dynamic>{
    'id': replyId,
    'sender': replySender,
    'snippet': replySnippet,
  };
  return '@@reply:${jsonEncode(meta)}@@\n$body';
}

File: frontend/src/pages/chat/utils/replyCodec.js
export function decodeReply(raw = '') {
    if (typeof raw !== 'string') return { reply: null, body: '' };
    const prefix = '@@reply:';
    if (!raw.startsWith(prefix)) return { reply: null, body: raw };
    const end = raw.indexOf('@@', prefix.length);
    if (end === -1) return { reply: null, body: raw };
    const jsonPart = raw.slice(prefix.length, end);
    const rest = raw.slice(end + 2);
    const body = rest.startsWith('\n') ? rest.slice(1) : rest;
    try {
        const reply = JSON.parse(jsonPart);
        if (!reply || typeof reply !== 'object') return { reply: null, body: raw };
        return { reply, body };
    } catch {
        return { reply: null, body: raw };
    }
}

export function stripReply(raw = '') {
    return decodeReply(raw).body;
}

export function encodeReply(body, replyTo) {
    const cleanBody = (body ?? '').toString();
    if (!replyTo) return cleanBody;
    const meta = {
        id: replyTo.id,
        sender: replyTo.sender,
        snippet: replyTo.snippet,
    };
    return `@@reply:${JSON.stringify(meta)}@@\n${cleanBody}`;
}

